dace/codegen/targets/fpga.py

```diff
@@ -690,12 +690,68 @@ def generate_scope(self, sdfg, dfg_scope, state_id, function_stream,
                                            callsite_stream,
                                            skip_entry_node=True)
 
-    def allocate_array(self, sdfg, dfg, state_id, node, function_stream,
-                       declaration_stream, allocation_stream):
+    def declare_array(self, sdfg, dfg, state_id, node, nodedesc,
+                      function_stream, declaration_stream):
+
+        # NOTE: We currently only support Arrays (not Views)
+        # that are dependent on non-free SDFG symbols.
+        assert (isinstance(nodedesc, dt.Array)
+                and not isinstance(nodedesc, dt.View) and any(
+                    str(s) not in sdfg.free_symbols.union(sdfg.constants.keys())
+                    for s in nodedesc.free_symbols))
+
+        result_decl = StringIO()
+        arrsize = nodedesc.total_size
+        dataname = node.data
+
+        # Check if array is already declared
+        try:
+            self._dispatcher.declared_arrays.get(dataname)
+            return  # Array was already declared in this or upper scopes
+        except KeyError:  # Array not declared yet
+            pass
+
+        allocname = cpp.ptr(dataname, nodedesc, sdfg)
+
+        if nodedesc.storage == dtypes.StorageType.FPGA_Global:
+
+            if self._in_device_code:
+
+                # TODO: Is this needed here?
```

Comment:
We added this safety check to verify that all global arrays (even in nested SDFGs) have been allocated before we reach this point (we are in device code). I would leave it, at least for the moment being